### PR TITLE
Changed default NGrams for FeaturizerText from 1 to 2

### LINKED ISSUE #4749: [Meta Issue] Changing defaults
First of all, is this a breaking change? @terrajobst suggested it is an acceptable breaking change https://github.com/dotnet/machinelearning/issues/2305#issuecomment-508197927, @eerhardt commented that is was decided that changing defaults is acceptable https://github.com/dotnet/machinelearning/issues/2305#issuecomment-509289908

- Change the following defaults based on results from @justinormont:

  - #2870 `FeatirizeText` default n-grams lengths to match default text recipe:
    - word: 1 -> 2
    - char: no change
    - @justinormont correct me if I am wrong: was the default weighting TF-IDF in TLC? It is currently TF

  - #2305 `AveragedPerceptron` default iterations from 1 to 10

- #3365 `RowGroupColumnName` in `FastTree` and `LightGbm` ranking trainer `Options` classes: `null` -> `'GroupId'`
  - This won't actually break anything since a ranking trainer would throw anyway if there was no group ID column specified https://github.com/dotnet/machinelearning/issues/3365#issuecomment-497775736
  - This would make it consistent with the API for the rankers without the `Options` class, where the default is `'GroupId'`
  - They should be consistent, at least. Either `null` everywhere or `'GroupId'` everywhere.

cc: @harishsk @justinormont @gvashishtha @eerhardt

### PR DESCRIPTION
**Pull request by michaelgsharp**
This fixes #4749. All the other changes in that issue have been done, this was the only one that was left, which will now allow us to close this issue.

@justinormont can you just verify that is the correct NGramLength to change? Based on the issue, and the other linked issue, this is what I believe is the correct spot. The standard error does go down in the tests with this, which is the expected behavior based on the linked issues.

### CODE REVIEW

**Comment by harishsk** [comment]
The issue #4749 is a tracking issue for many changes. Once this PR is merged, github is going to automatically close that issue because of the text "fixes #..." in the description above. 
Can you please remove that text or add some other word after "fixes" so that this bug isn't automatically closed?

We can close this issue after all the issues that are tracked by #4749 are fixed.

**Comment by michaelgsharp** [comment]
Hey Harish, all the other issues in there have been fixed already, I was planning on changing them all in this PR, but going through them all they were already changed. I am not sure why the other PR's didn't reference that issue. Let me see if I can find the other PR's and link them to this one.

**Comment by michaelgsharp** [comment]
Actually, let me double check. There are several places that defaults are used, let me make sure I changed them all. If I missed some I will add those changes to this PR as well.

File: test/BaselineOutput/Common/Text/featurized.tsv [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
...d.power	Word.power|abuse,	Word.abuse,	Word.abuse,|this	Word.this|*really*	Word.*really*	Word.*really*|came	Word.came|as	Word.a|surprise	Word.surprise	Word.surprise|to	Word.me|when	Word.when|someone	Word.someone|e-mailed	Word.e-mailed	Word.e-mailed|this	Word.this|info	Word.info|to	Word.this|morning!	Word.morning!	Word.morning!|sorry	Word.sorry|he	Word.he|couldn't	Word.couldn't	Word.couldn't|be	Word.be|more	Word.more|adult	Word.adult	Word.adult|with	Word.with|his	Word.his|admin	Word.admin|powers,	Word.powers,	Word.powers,|but	Word.but|as	Word.as|stan	Word.stan	Word.stan|lee	Word.lee	Word.lee|said	Word.said|over	Word.over|four	Word.four|decades	Word.decades	Word.decades|ago,	Word.ago,	Word.ago,|with	Word.with|great	Word.great	Word.great|power	Word.power|comes	Word.comes|great	Word.great|responsibility.	Word.responsibility.	Word.responsibility.|of	Word.of|course,	Word.course,	Word.course,|the	Word.the|big	Word.big	Word.big|question	Word.question	Word.question|now	Word.now|is	Word.is|who	Word.who|matthew	Word.matthew	Word.matthew|fenton	Word.fenton	Word.fenton|will	Word.will|run	Word.run	Word.run|and	Word.and|hide	Word.hide	Word.hide|behind	Word.behind	Word.behind|when	Word.when|he	Word
```
Comment (justinormont) [inline test/BaselineOutput/Common/Text/featurized.tsv:7]: Bigrams are in the output now, which is good. For example, `Word.generally|accepted`. Tri-charactergrams are present.

The change from unigrams to bigrams appears to be working.

File: src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs [hunk 2]
```diff
@@ -61,7 +61,7 @@ public abstract class OnlineLinearOptions : TrainerInputBaseWithLabel
         [BestFriend]
         internal class OnlineDefault
         {
-            public const int NumberOfIterations = 1;
+            public const int NumberOfIterations = 10;
```
Comment (justinormont) [inline src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: Two smaller PRs will be easier to review. 

Together, this is modifying 105 files by changing both the default ngram length, and AveragedPerceptron's default iterations.

For instance, the AveragedPerceptron change should only affect baseline outputs where AveragedPerceptron is used (including in ensembles). But I see some changes in SVM files, which shouldn't change due to the AP change, but might have a TextTransform in the pipeline. It is just hard to disentangle.

File: test/BaselineOutput/Common/FeatureContribution/SVMBinary.tsv [hunk 3]
```diff
@@ -12,7 +12,7 @@
 #@   col=FeatureContributions:R4:32-37
 #@   col=FeatureContributions:R4:38-43
 #@ }
-950	757	692	720	297	7515	1	1	950	757	692	720	297	7515	0.956696868	0.760804	0.7872582	0.754716933	0.297893673	0.7578661	0	0.05771441	0	0	0.0166732743	1	18	0:-0.99999994	2:-0.838546038	6:-0.111203589	11:1	12:-0.688422263	17:6.190648
-459	961	0	659	274	2147	0	0	459	961	0	659	274	2147	0.462235659	0.965829134	0	0.690775633	0.27482447	0.21651876	0	0.256453544	0	0	0.053840857	1	18	0:-1	6:-0.188063636	11:1	12:-0.332616657	17:1.76863885
-672	275	0	65	195	9818	1	1	672	275	0	65	195	9818	0.6767372	0.2763819	0	0.06813417	0.195586756	0.990116954	0	0.0160482265	0	0	0.008379247	1	18	0:-0.99999994	6:-0.060210254	11:1	12:-0.4869682	17:8.087795
-186	301	0	681	526	1456	0	0	186	301	0	681	526	1456	0.187311172	0.302512556	0	0.713836432	0.527582765	0.1468334	0	0.118446559	0	0	0.15241152	1	18	0:-1	6:-0.112376556	11:1	12:-0.134785831	17:1.19941235
+950	757	692	720	297	7515	1	1	950	757	692	720	297	7515	0.956696868	0.760804	0.7872582	0.754716933	0.297893673	0.7578661	0.239822984	0	0	0.0266424417	0	1	18	2:-1	8:-0.08292028	11:1	14:-0.6742478	17:8.131277
```
Comment (justinormont) [inline test/BaselineOutput/Common/FeatureContribution/SVMBinary.tsv:15]: I wouldn't anticipate this changing due to changing AveragedPerceptron's iterations from 1 to 10. Is there a TextTransform in this pipeline causing the change?

File: src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs (hunk 2 above)
Comment (justinormont) [inline src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: Modifying within `OnlineLinear.cs` seems to be affecting multiple trainers besides AveragedPerceptron. You may need to find a way to target just AveragedPerceptron.

Comment (michaelgsharp) [reply src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: Yeah, I am going to be breaking this up. We seem to have a memory issue with the NGram stuff, so I will be splitting these out  to get these in while I work on that.

You said the SVM files shoudln't change, but might have a TextTransform in the pipeline. If they do have a text transform should they change? Or should we only see changes where AveragedPerceptron is used and no where else? Or, is this change a bad thing for svm as well?

Comment (michaelgsharp) [reply src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: I can do that. Question though, would it be bad to update it for them all? Or is just that we know for AveragedPerceptron 10 is better, but thats not true for the others?

Comment (justinormont) [reply src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: ### Expected output changes 
This PR is changing both AveragedPerceptron and the text transform. If the pipeline has either of these components, I would expect the pipeline's output to change.

Any pipeline with neither AveragedPerceptron nor the text transform, should have the same output as before. Modifying the other trainers' iteration counts is not benchmarked and, by default, I would assume will not be beneficial.

### Memory footprint
The memory footprint will increase and is absolutely worth it for the accuracy gains of our default pipeline. The pipeline being implemented in this PR, Bigrams+tricharactergrams w/ AveragedPerceptron{iter=10}, is the default in TLC and ML․NET AutoML for text and is highly tested/benchmarked. It's the result of around 50-150k runs on the combinatorics of a variety of datasets; primarily testing for ngram lengths, trainers, and hyperparameters. This pipeline was found to be an optimal default. Other featurization options help on specific datasets (see: [list](https://github.com/dotnet/machinelearning-modelbuilder/issues/702#issuecomment-628283455)).

The memory increase will occur in a few places, (1) in the text transform's dictionary where it stores all seen ngrams, (2) in the iDataView cache if used, (3) in tree trainers at the dataset transpose step for row-wise to column-wise, and (4) in high multi-class linear trainers which store the weight for each feature times each class.

In the disentangled PR handing the ngram change, we should dig into any memory issues which arise. It may be a limitation of our build/testing platform.

Perhaps in the future we should write a doc on tips for low memory model training and serving.

Comment (michaelgsharp) [reply src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: Thanks for the detailed responses @justinormont.

I will check the files that were changed again and make sure that they all have either AveragedPerceptron or a text transform. If I find any that don't I will change the location where I am overwriting that value.

I think I have the memory increase issue figured out, but I am still looking into it. Yes, the memory should go up, but it was exploding way faster then it should. A 70 Kb file (250 lines) was exploding into 1.7 GB _per Ngram length_ during training. So when the NGram length was 2, it was using 3.4 GB for that small file. It appears to be because we have implemented it so it creates all the possible memory upfront, whether it is being used or not. I have changed that and the memory usage is much smaller now. Still verifying that my changes had no impact on any other tests though.

Comment (justinormont) [reply src/Microsoft.ML.StandardTrainers/Standard/Online/OnlineLinear.cs:64]: For the other trainers, it's unknown if 10 iterations will be better or worse. For most trainers, `NumberOfIterations`/`NumberOfTrees` and [`LearningRate`](https://en.wikipedia.org/wiki/Learning_rate) are the primary hyperparameters; changing them has a large impact.

Since the existing defaults in the other trainers are backed by some level of benchmarking when they were created, I would keep with the existing value. Moving from 1 to 10 iterations, we know the runtime will increase and we should demonstrate the value (accuracy gain) of this extra time spent.

For AveragedPerceptron, we did benchmarks showing the gain in having 10 iterations as the new default. 

Subset of the AveragedPerceptron benchmarks are in the original issue -- https://github.com/dotnet/machinelearning/issues/2305.

**Comment by michaelgsharp** [comment]
Closing this PR as #5248, #5258, #5290 fix this issue.
